animation_nodes/nodes/mesh/mesh_points_scatter.py

```diff
@@ -0,0 +1,106 @@
+    nodeSeed: IntProperty(update = propertyChanged)
+
+    triangulate: BoolProperty(name = "Triangulate Mesh", default = False,
+        update = AnimationNode.refresh)
+
+    def setup(self):
+        self.randomizeNodeSeed()
+
+    def create(self):
+        self.newInput("Mesh", "Mesh", "mesh", dataIsModified = True)
+        self.newInput("Integer", "Seed", "seed")
+        self.newInput("Integer", "Amount", "amount", value = 10, minValue = 0)
+        self.newInput("Float List", "Weights", "weights")
+
+        self.newOutput("Vector List", "Points", "points")
+
+    def draw(self, layout):
+        row = layout.row(align = True)
+        row.prop(self, "nodeSeed", text = "Node Seed")
+        row = layout.row(align = True)
+        row.prop(self, "triangulate", text = "Triangulate Mesh")
+
+    def execute(self, mesh, seed, amount, weights):
+        vertices = mesh.vertices
+        polygons = mesh.polygons
+
+        if len(vertices) == 0 or len(polygons) == 0 or amount == 0:
+            return Vector3DList()
+
+        if self.triangulate: polygons = triangulatePolygons(polygons)
+        if polygons.polyLengths.getMaxValue() > 3:
+            self.raiseErrorMessage("Mesh should have triangle polygons.")
+
+        return self.randomPointsScatter(vertices, polygons, seed, amount, weights)
+
+    # Mesh Sampling basis on Barycentric Coordinate System: http://www.joesfer.com/?p=84
+    def randomPointsScatter(self, vertices, polygons, seed, amount, weights):
+        areas = trianglePolygonsArea(vertices, polygons)
+        areaMin = areas.getMinValue()
+        if areaMin == 0: self.raiseErrorMessage("Remove doubles.")
+
+        if len(weights) != 0:
+            usePolyWeight = True
+            polyAmount = len(polygons)
+            polyWeights = FloatList(length = polyAmount)
+            weights = VirtualDoubleList.create(weights, 1).materialize(polyAmount)
+            for i, polygon in enumerate(polygons):
+                polyWeights[i] = (weights[polygon[0]] + weights[polygon[1]] + weights[polygon[2]]) / 3
+        else:
+            usePolyWeight = False
+
+        probabilities = LongList()
+        if usePolyWeight:
+            for i, area in enumerate(areas):
+                polyRange = int(area * polyWeights[i] / areaMin)
+                for j in range(polyRange):
+                    probabilities.append(i)
+        else:
+            for i, area in enumerate(areas):
+                polyRange = int(area / areaMin)
+                for j in range(polyRange):
+                    probabilities.append(i)
+
+        lenProb = len(probabilities)
+        if lenProb == 0: return Vector3DList()
+        points = Vector3DList(length = amount)
+        for i in range(amount):
+            randIndex = probabilities[int(uniformRandomDoubleWithTwoSeeds(seed + i, self.nodeSeed, 0, lenProb))]
+            polygon = polygons[randIndex]
+            points[i] = self.randomPolyPoint(i, vertices, polygon, seed)
+        return points
```

Comment:
One problem with this approach is that you will often need to go to memory to get the triangle. An alternative better algorithm will put caching into consideration and generate all points of a single triangle successively.

Reply:
Omar, we do not know prior to the total number of points on a polygon.

1. We can cache polygon (or its vertices by polygon index) then use it for the points generations.
2. Or, we split this loop into two: first, calculate how many points each polygon has then in the second loop generates those points.

What do you think?

Reply:
I don't think (1) would be very useful in this case.
We can try (2). The best thing you can do is time all of this. Time all the methods you can think of and see which one is best.

Reply:
Omar, with the method(2), I have implemented the main algorithum in the Cython.